### PR TITLE
Add default update action to actions dropdown in object detail modal

### LINKED ISSUE #32322: Add default update action to actions dropdown in object detail modal
Epic: #32320
Product doc: [link](https://www.notion.so/Allow-basic-deletes-and-updates-on-object-detail-views-f05b298c3aba4a6ca883b2738d86824d)
Depends on: #32321
Related to: #32323

----

- Clicking update should open the existing basic action update form on top of the detail modal. The user can make whatever edits they wish and hit update. The update modal should close and the user should see any updated values present in the detail modal along with a confirmation toast.

### PR DESCRIPTION
**Pull request by kamilmielnik**
Closes #32322

### Description

- [Hide whitespace](https://github.com/metabase/metabase/assets/6830683/cde5ae72-2216-4ddd-911c-68ef00bd7f70) is recommended when reviewing this PR
- Tests will be added in #32350

### How to verify

> Clicking update should open the existing basic action update form on top of the detail modal. The user can make whatever edits they wish and hit update. The update modal should close and the user should see any updated values present in the detail modal along with a confirmation toast.

### Demo

https://github.com/metabase/metabase/assets/6830683/97d990d2-e475-42e3-8397-235bf7ff0a6a

### CODE REVIEW

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx
Comment (kamilmielnik) [inline frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx:462]: Reason for this line change: https://github.com/metabase/metabase/pull/32331#discussion_r1267827835

File: frontend/src/metabase/dashboard/actions/actions.ts
Comment (kamilmielnik) [inline frontend/src/metabase/dashboard/actions/actions.ts:74]: What happens here depends on where the action was triggered from (e.g. which screen), so I moved it into the component where the action is called.

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx
Comment (kamilmielnik) [inline frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:71]: Preserves the old behavior (`.catch(_.noop)`).

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx
Comment (kamilmielnik) [inline frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:40]: I refactored `ActionParametersInputForm` so that it no longer depends on `dashboard` and `dashcard`. This way it can be reused in different contexts.

File: frontend/src/metabase/actions/components/ActionViz/ActionVizForm.tsx
Comment (kamilmielnik) [inline frontend/src/metabase/actions/components/ActionViz/ActionVizForm.tsx:94]: I have no idea why it's here, but I preserved the current behavior.

**Review by cdeweyx — APPROVED** [review]
Functionality looks good. Will give more detailed feedback (nitpicks etc.) before merging to master.

**Review by kdoh — APPROVED** [review]
Works as I'd expect on the update mechanics front. There biggest thing I'm running into is that the object detail modal seems to re-fetch after the update and I was getting some layout flashing as a result. In my case it might've been because the model I was testing this on has 'articles' pointing back at it, so I have the `connected to` side panel in my detail view.

Perhaps some data fetching stuff to work out?

**Comment by kamilmielnik** [comment]
> Perhaps some data fetching stuff to work out?

Correct. @kdoh FYI The problem is there not only when you update the object, but also when you use the up/down arrows to navigate between subsequent objects in the modal ([video](https://github.com/metabase/metabase/assets/6830683/93a5816f-dea3-4ef1-95ae-4167e507327b)).

File: frontend/src/metabase/actions/components/ActionViz/ActionVizForm.tsx
Comment (uladzimirdev) [inline frontend/src/metabase/actions/components/ActionViz/ActionVizForm.tsx:108]: there are two other places with the same check. probably it's a sign for having a helper

File: frontend/src/metabase/actions/containers/ActionExecuteModal/ActionExecuteModal.tsx
Comment (uladzimirdev) [inline frontend/src/metabase/actions/containers/ActionExecuteModal/ActionExecuteModal.tsx:53]: nit: with modern js you can have `onClose?.()`, which allows you do have `onClose` not defined

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx
Comment (uladzimirdev) [inline frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:71]: it would be nice to have it `console.error`ed IMO

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx
Comment (uladzimirdev) [inline frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:112]: nit: `!!` is shorter

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx
Comment (uladzimirdev) [inline frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx:338]: do you really need `useCallback` for modal actions?

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx
Comment (uladzimirdev) [inline frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx:413]: it's either a number or undefined, seems `!!executeActionId` would be enough

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx:38]: good point, I'm fine with this change, but I believe the proper place for this auto-generating ID is mockHelpers

**Review by uladzimirdev — COMMENTED** [review]
LGTM overall, I'm not quite sure I got all the prop renamings, but if it provides more context, 👍 

left a couple of comments

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx]: :+1: 357d720eea143b4263c84f7d90a114a05536202b

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx:413]: Your suggestion does not handle `0` correctly.

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:71]: Will it cause the error to be logged somewhere in our systems?

I don't really like to console errors out - I'm a user of some banking systems that spill lots of errors into JS console and it gives me the shivers every time I see it.

BTW this is existing code, just refactored to async/await + try/catch. The error is currently [ignored silently in `master`.](https://github.com/metabase/metabase/blob/73966fb/frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx#L84)

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:112]: Shorter does not mean better or worse - it depends.

FYI I stopped using this notation few years ago for these reasons:
1. readability - if a line has more operators than a single boolean cast (`!!`), then it gets more difficult to read
2. verbosity - I've seen `!!` that were actual typos
3. consistency - cast types only in one way throughout the codebase
    - with all boolean casts (you're probably gonna want to `array.filter(Boolean)`, not `array.filter(item => !!item)`
    - with other types casts (I think only casting to `Number` has a similar "trick", e.g. `+"7"`)
4. when working in a team (or on an open source project), people not familiar JS/TS might come and read the code - `!!` is definitely more cryptic than `Boolean()`

I'm fine with this notation (`!!`) being used as long as the code is still readable (see point 1), but unless it's explicitly forbidden I will continue to cast types using type constructors as functions.

File: frontend/src/metabase/actions/containers/ActionExecuteModal/ActionExecuteModal.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/actions/containers/ActionExecuteModal/ActionExecuteModal.tsx]: 6eca96e23975a56ab12d85f163198060f07b0934

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx]: I moved it to `__support__/utils.ts`. It's a new file, it didn't make sense to put it in any of the existing ones.

2a6c0d69bf44bb255d9ead0bfda4c315dad35bab

File: frontend/src/metabase/actions/components/ActionViz/ActionVizForm.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/actions/components/ActionViz/ActionVizForm.tsx]: 86d158832b84fbca158158ea8fa1f6427e15d693 + 6c568b904cd6a2bba738ee3fbd925e59ce976d0c

File: frontend/src/metabase/dashboard/actions/actions.ts
Comment (WiNloSt) [reply frontend/src/metabase/dashboard/actions/actions.ts:74]: I asked questions in `frontend/src/metabase/actions/components/ActionViz/Action.tsx`. I'm not sure why this should be lifted up.

File: frontend/src/metabase/visualizations/components/ObjectDetail/utils.ts
Comment (WiNloSt) [inline frontend/src/metabase/visualizations/components/ObjectDetail/utils.ts:157]: I don't think this is correct. If you make either edit or delete action public. You won't be able to use those action in the model detail anymore.

Screenshot 2023-07-24 at 6 31 01 PM

image

**Comment by WiNloSt** [comment]
The above review is only for actual code review. I'll review tests later as I need a break.

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:71]: I totally understand it, @ranquild what's our policy regarding logging errors?

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:112]: as for filter, I'd use Boolean for visibility, as for this specific case, I'd use `!!` - just personal preference. But it's `nit`

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx
Comment (uladzimirdev) [reply frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx:413]: you're right, but it is a real use case?

**Review by uladzimirdev — APPROVED** [review]

**Comment by uladzimirdev** [comment]
@kamilmielnik please wait for @WiNloSt to finish his review

**Review by WiNloSt — CHANGES_REQUESTED** [review]
Block until this [comment is addressed or answered](https://github.com/metabase/metabase/pull/32433#discussion_r1272139034) to prevent accidental merge.

File: frontend/src/metabase/actions/components/ActionViz/Action.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/actions/components/ActionViz/Action.tsx:111]: 1. I found there's no reason to `reloadDashboardCards()` if the result is not successful. It's an optimization really - I'm avoiding a redundant API call here.
2. [This comment](https://github.com/metabase/metabase/pull/32433#discussion_r1269402320) explains it. There is no difference in how it works - it's a matter of "what" (e.g. which component/action/thunk/etc.) is responsible for this. I wanted to be consistent with `dispatch(runQuestionQuery())` call in `ObjectDetail` - notice it's `ObjectDetail` that decides what data to reload when action is successful - this way `ActionExecuteModal` is unaware of where it's used. I wanted to make `executeRowAction` unaware of where it's used. Please let me know if you disagree with this change - it's not required in this PR.

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx:413]: I don't want a future reader to look at this code and wonder why 0 is handled differently than all the other numbers - it looks like a bug.

File: frontend/src/metabase/actions/containers/ActionExecuteModal/ActionExecuteModal.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/actions/containers/ActionExecuteModal/ActionExecuteModal.tsx:44]: > Why do you remove memoization from this callback? It seems it's being used as another hook's dependency

It's `onSubmitSuccessRef`, not `onSubmitSuccess`.

File: frontend/src/metabase/visualizations/components/ObjectDetail/utils.ts
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/components/ObjectDetail/utils.ts:157]: > I don't think this is correct. If you make either edit or delete action public. You won't be able to use those action in the model detail anymore.

Public actions require a separate endpoint and it's been put on hold for now, please see https://metaboat.slack.com/archives/C057T1QTB3L/p1689801698386559?thread_ts=1689665950.493399&cid=C057T1QTB3L

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:38]: Any prop that is referenced in hooks (callback or effect) after the first `await` statement is prone to suffer from stale closure values. This can result in nasty bugs. 

This change also prevents bugs in future usages of this component.

There are few articles about it, here's one: https://css-tricks.com/dealing-with-stale-props-and-states-in-reacts-functional-components/

**Comment by kamilmielnik** [comment]
> I'll review tests later as I need a break.

@WiNloSt PR with tests is a WIP, please hold on with reviewing it.

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:93]: Sorry, I don't understand. Can you please rephrase?

The condition with keys count has been moved to `ActionVizForm` where `dashboard` and `dashcard` are always present.
The point of refactoring `ActionParametersInputForm` was to make it unaware of `dashcard` and `dashboard`.

File: frontend/src/metabase/actions/components/ActionViz/ActionVizForm.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/actions/components/ActionViz/ActionVizForm.tsx:91]: You're absolutely right, `dashcardParamValues` should not be renamed in this component :+1: Good catch.

299f0aec17b8f7400c88cc383271ec442db861b8

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx
Comment (ranquild) [reply frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:71]: I don't know if such a policy exists; however browser logs are usually attached to issues in GitHub

Comment (kamilmielnik) [reply frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx]: > however browser logs are usually attached to issues in GitHub

It makes sense to add it then. Done: 7b2507fc2d09ca8ef55e47d82a050d18fcd98256

File: frontend/src/metabase/actions/containers/ActionExecuteModal/ActionExecuteModal.tsx
Comment (WiNloSt) [reply frontend/src/metabase/actions/containers/ActionExecuteModal/ActionExecuteModal.tsx:44]: Ah you're right. That's a bit hard to spot.

File: frontend/src/metabase/actions/components/ActionViz/Action.tsx
Comment (WiNloSt) [reply frontend/src/metabase/actions/components/ActionViz/Action.tsx:111]: Re: 2. This makes more sense. I'm all for consistency 👍

File: frontend/src/metabase/visualizations/components/ObjectDetail/utils.ts
Comment (WiNloSt) [reply frontend/src/metabase/visualizations/components/ObjectDetail/utils.ts:157]: I think I understand that in a different way.

I think:
1. A normal model should always show edit/delete if actions are enabled, basic actions are enabled.
2. A model can be put on a dashboard, though I couldn't find a way to open object detail from models on a dashboard.
3. A model action could be made public. But that shouldn't take away the ability to edit/delete when not viewed publicly (point 1)
4. I think if we have a way to view object detail publicly e.g. via public dashboards then that's what is not included in the scope of the project.

cc @cdeweyx for clarification.

Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/components/ObjectDetail/utils.ts:157]: > 2. A model can be put on a dashboard, though I couldn't find a way to open object detail from models on a dashboard.

Yeah, it's not possible, see [product doc](https://www.notion.so/metabase/Allow-updates-and-deletes-on-model-object-detail-views-f05b298c3aba4a6ca883b2738d86824d?pvs=4#75d2dac5e2c7477faba202b6e803d922).

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx
Comment (WiNloSt) [reply frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:38]: I thought you encountered some bugs already. In this case, I think both `onSubmitSuccess` and `prefetchValues` (which depends on `fetchInitialValues`) should be safe to be used directly because:
1. `onSubmitSuccess` just call `onClose` and `onSuccess` which calls these functions that don't have any dependencies on them.
    - https://github.com/metabase/metabase/blob/6c568b904cd6a2bba738ee3fbd925e59ce976d0c/frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx#L338-L340
    - https://github.com/metabase/metabase/blob/6c568b904cd6a2bba738ee3fbd925e59ce976d0c/frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx#L165-L167
2. From the usage of `prefetchValues` it feels wrong that we could potentially use a newer thus different `prefetchValues` with older `onSubmit`. e.g. what is `onSubmit` is for one model but somehow `perfetchValues` is for another model since we've somehow navigate to the other model. 
    https://github.com/metabase/metabase/blob/6c568b904cd6a2bba738ee3fbd925e59ce976d0c/frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx#L87-L93

This kind of interaction seems like you should use the callback that's defined for each render cycle. If there's anything needed to be updated it will just get updated via the `useEffect` or props change anyway. The example would be you wait for the API to finish, then you close the current modal. It shouldn't be that while you're waiting for the API to finish you close the modal then open another model detail and the API finishes then it calls the latest `onSubmitSuccess` which closes the new modal you just opened.

I think the use case is different from that article you mentioned. TBH, I'm not 100% right now so I couldn't say for sure this is wrong but I have a hunch that this sounds wrong. Please feel free to discuss this.

tl;dr: I'd not use `useLatest` since I feel it could cause bugs not preventing it.

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx
Comment (WiNloSt) [reply frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:93]: Ok that clears things up.

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx
Comment (WiNloSt) [inline frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx:85]: What do you think about leaving the implementation of `getNextId` in this file? I'm not sure it should be abstracted yet.

I'm not aware of other tests that need this functionality. If there are a lot of other tests that want to do this then it might warrant putting it in a util.

One thing I think cause trouble with our E2E utils is that we don't plan them well enough, creating util is easy, but making util coherent with the rest is not easy.

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx
Comment (WiNloSt) [inline frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx:1]: I don't see a test that actually renders the action form e.g. when you click either **update** or **delete**. Since this modal is added to `ObjectDetail` I think we should test that as well.

Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx:1]: This will be covered by e2e tests in #32350, also asserting that the update form gets prefilled properly.
Do you think this needs to be double-checked in unit tests?

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx:85]: Actually I already wanted to have this utility few weeks ago (can't remember the PR). 

> One thing I think cause trouble with our E2E utils is that we don't plan them well enough, creating util is easy, but making util coherent with the rest is not easy.

I will keep that in mind for the future. I believe this case is justified.

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx
Comment (WiNloSt) [reply frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx:1]: I think we should double-check in unit tests since it's the action form is added to ObjectDetail. It feels straightforward to test new subcomponents added to the component under test. It won't take long time either since Integration test (jest) is relatively fast compared to Cypress.

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx
Comment (WiNloSt) [reply frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx:85]: Sounds reasonable then.

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:38]: > 1. `onSubmitSuccess` just call `onClose` and `onSuccess` which calls these functions that don't have any dependencies on them.

That was one of my points - the consumer of this component should not worry about this.
You're right - with the current state of the code it does not matter. But consumers of this component can change in the future and suddenly there might be a bug in the child component, that wasn't modified.

> 2. From the usage of `prefetchValues` it feels wrong that we could potentially use a newer thus different `prefetchValues` with older `onSubmit`. e.g. what is `onSubmit` is for one model but somehow `perfetchValues` is for another model since we've somehow navigate to the other model. 

It's the other way around - we need to use the latest `prefetchValues` because `prefetchValues` is possibly for another model by then.

> Please feel free to discuss this.

Perhaps it'd be most efficient to discuss this on a call.

File: frontend/src/metabase/visualizations/components/ObjectDetail/utils.ts
Comment (cdeweyx) [reply frontend/src/metabase/visualizations/components/ObjectDetail/utils.ts:157]: What Kamil said

**Review by WiNloSt — APPROVED** [review]
So I and Kamil had a call, and he's gonna revert the `useLatest` change. Though, there are always counter-examples for both of our arguments so we are both right.

We agreed on reverting `useLatest` back for consistency because there isn't really `useLatest` usage in our project.

It's worth noting that he shared his experience with using `useLatest` to handle stale props works more often than not. Which I think is something to keep in mind.

File: frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx:38]: We had a call about it, here are notes and conclusions:
- it seems that both approaches are valid with the current state of code in this component and how it's used
- we think both approaches can be counter-exampled and shown to not work as expected (in some cases)
- we decided not to use `useLatest` for now because it hasn't been used in the code extensively (be consistent)
- we think it's good to talk this problem through during frontend guild meeting
  - we should have aforementioned counter-examples provided in an RFC
  - we don't know what the conclusion of that RFC would be

Comment (kamilmielnik) [reply frontend/src/metabase/actions/containers/ActionParametersInputForm/ActionParametersInputForm.tsx]: Done: 9b779f0a28f393d4017c671d777621aa4ccb33ae

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx:1]: > I think we should double-check in unit tests since it's the action form is added to ObjectDetail. It feels straightforward to test new subcomponents added to the component under test. It won't take long time either since Integration test (jest) is relatively fast compared to Cypress.

I'm still thinking whether it's worth doing (probably yes). Anyway, that's for another PR: #32350.

Comment (WiNloSt) [reply frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx:1]: Yeah, don't need to be in this PR. So it won't block this PR from getting merged.

File: frontend/src/metabase/actions/components/ActionViz/ActionVizForm.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/actions/components/ActionViz/ActionVizForm.tsx:91]: [Aforementioned commit](https://github.com/metabase/metabase/commit/299f0aec17b8f7400c88cc383271ec442db861b8) went into wrong branch, cherry-picked into the right one: db98b2b0fe

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx
Comment (kamilmielnik) [reply frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.unit.spec.tsx:1]: > I don't see a test that actually renders the action form e.g. when you click either **update** or **delete**. Since this modal is added to `ObjectDetail` I think we should test that as well.

d6d0a39cc044866870137e6ea94b08049aa8a017 + 2cfe7b3b103ca37b417db9ea9f8e4a6cb6621db3 + 6bdfb3cccaa4f014c74136db2466efcb96d357b4

**Pull request merged**
